Custom storage for SCORM Xblock

naumanshafi opened the pull request:
### PR Description

**Summary:**
Implement a custom storage solution to store and retrieve data from a private S3 bucket. Modify the handling of SCORM content requests loaded via JavaScript from a primary file, ensuring they are routed through a custom XBlock view instead of directly accessing the S3 storage. All SCORM XBlock resource requests should be intercepted by the custom view to add signatures before forwarding them to the S3 storage, allowing secure access to private content. Additionally, the ability to read SCORM content from a private bucket should be configurable via Django settings.

**Changes Made:**
- Created a custom storage implementation to facilitate data storage in a private S3 bucket.
- Implemented a custom XBlock view to handle SCORM content requests.
- Modified the routing of SCORM XBlock resource requests to pass through the custom view for signature addition.
- Added configuration options in Django settings to control the reading of SCORM content from a private bucket.

**Motivation:**
This PR addresses [EDE-2041](https://edlyio.atlassian.net/browse/EDE-2041), the Jira issue discussing the need for secure storage and retrieval of SCORM content. By leveraging a private S3 bucket and introducing a custom XBlock view, we ensure that SCORM content requests are processed securely with the addition of signatures. The configurable settings enable flexibility in managing access to SCORM content stored in private buckets.

**Additional Notes:**
- The custom storage solution follows best practices for secure data handling and authentication.
- The custom XBlock view integrates seamlessly into the existing SCORM content loading process.
- The provided Django settings allow for easy configuration of the SCORM content retrieval from private buckets.

### Related Issues
- [EDE-2041](https://edlyio.atlassian.net/browse/EDE-2041): Discusses the need for secure storage and handling of SCORM content.

### Configuration Changes (lms/envs/private.py and cms/envs/private.py)
```python

# XBlock settings for ScormXBlock
XBLOCK_SETTINGS["ScormXBlock"] = {
    "STORAGE_FUNC": "openedxscorm.storage.scorm_storage",
    "SCORM_S3_BUCKET_NAME": "<your-bucket-name>"
}
```

[EDE-2041]: https://edlyio.atlassian.net/browse/EDE-2041?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

ziafazal commented:
Hi @regisb could you please review this one?

regisb commented on openedxscorm/scormxblock.py:
```diff
@@ -202,6 +204,35 @@ def student_view(self, context=None):
+        path = re.sub(r"(\?.*|:[\d:]*$)", "", path)
+        file_name = os.path.basename(path)
+        signed_url = self.storage.url(os.path.join(self.extract_folder_path, path))
+        file_content = requests.get(signed_url).content
```
This is going to blow up the server memory if the object stored at the signed url is a large file. Can we issue a 302 redirect instead?

regisb commented on openedxscorm/scormxblock.py:
```diff
@@ -202,6 +204,35 @@ def student_view(self, context=None):
+        -------
+        Response object containing the content of the requested file with the appropriate content type.
+        """
+        path = request.url.split('scorm_view/')[-1]
```
I think we could use `_suffix` for that, right?

regisb commented on setup.py:
```diff
@@ -26,7 +26,7 @@ def package_data(pkg, roots):
 setup(
     name="openedx-scorm-xblock",
-    version="15.0.0",
+    version="15.1.0",
```
This PR introduces a backward-compatible change, so we should only bump to 15.0.1 https://docs.tutor.overhang.io/tutor.html#versioning

regisb commented on openedxscorm/scormxblock.py:
```diff
@@ -202,6 +204,35 @@ def student_view(self, context=None):
+        Response object containing the content of the requested file with the appropriate content type.
+        """
+        path = request.url.split('scorm_view/')[-1]
+        path = re.sub(r"(\?.*|:[\d:]*$)", "", path)
```
This is not the right way to parse a url. Use `urllib.parse` instead.

regisb commented on openedxscorm/scormxblock.py:
```diff
@@ -202,6 +204,35 @@ def student_view(self, context=None):
         return frag
 
+    @XBlock.handler
+    def scorm_view(self, request, _suffix):
```
As this is a proxy view, we should rename the function to `def assets_proxy`.

regisb commented on openedxscorm/scormxblock.py:
```diff
@@ -331,15 +362,28 @@ def extract_package(self, package_file):
+                folder = self.extract_folder_base_path
+                logger.warning("Serving SCORM content from old-style path: %s", folder)
+            return self.storage.url(os.path.join(folder, self.index_page_path))
+        else:
```
The XBlock should not have to be aware of all this logic. Instead, this function should not be modified and all the logic should be moved to the `self.storage.url` function. Since `self.storage` will be an instance of `S3ScormStorage` it will be possible to override the `url` method there.

regisb commented on openedxscorm/storage.py:
```diff
@@ -0,0 +1,37 @@
+                         querystring_expire=querystring_expire)
+
+
+def scorm_storage(xblock):
```
This function is in the openedxscorm.storage module, so we wan rename it to simply "s3".

```
XBLOCK_SETTINGS["ScormXBlock"] = {
    "STORAGE_FUNC": "openedxscorm.storage.s3"
}
```

regisb commented on openedxscorm/storage.py:
```diff
@@ -0,0 +1,37 @@
+    """
+    bucket = xblock.xblock_settings.get('SCORM_S3_BUCKET_NAME', None)
+    querystring_auth = xblock.xblock_settings.get('SCORM_S3_QUERY_AUTH', None)
+    querystring_expire = xblock.xblock_settings.get('SCORM_S3_EXPIRES_IN', 604800)
```
There should be better defaults for these settings, such that a user Scorm assets are stored in the default S3 bucket if they are undefined. If possible, users should not have to define any extra scorm-specific setting. 

Also, these new settings should be properly documented in the README.

regisb reviewed (changes requested):
Thanks for the PR! I have some important comments here. Please see them below. In particular, the setup instructions that are currently in the PR description must be added to the xblock README to make it very clear to the end users how to make use of the new S3 storage backend.

naumanshafi commented:
I acknowledge the concerns you raised. If large files are fetched using the current approach, it could indeed lead to memory issues on the server. However, if we switch to issuing a 302 redirect, it would expose the signed URL in the network tab of the user's browser, which can lead to potential security risks.

Additionally, another challenge arises when serving HTML files that include relative links. After a 302 redirect, the base URL changes to the S3 bucket's URL, so the relative links within the HTML file will fail to load the resources as expected.

naumanshafi commented:
I understand the confusion here. but in this scenario, runtime is typically associated with an XBlock instance, not the storage class. Therefore, it is difficult to make the runtime.handler_url method was available in S3ScormStorage.

regisb commented:
> I understand the confusion here. but in this scenario, runtime is typically associated with an XBlock instance, not the storage class. Therefore, it is difficult to make the runtime.handler_url method was available in S3ScormStorage.

Right. But notice that the storage instance is created by the `scorm_storage` function, which does take the xblock as a first argument. Thus, the object can be passed to the storage class constructor.

What I am trying to avoid here is to add s3-specific code to the core.

naumanshafi commented:
Thank you for your thorough review and feedback on the PR. I have made all the requested changes and addressed the concerns raised. I kindly request you to review the PR again to ensure that the modifications meet the requirements and address your feedback appropriately.

regisb commented on README.rst:
```diff
@@ -82,12 +82,28 @@ By default, static assets are stored in the default Django storage backend. To o
-        "STORAGE_FUNC": "my.custom.storage.module.get_scorm_storage_function",
+        "STORAGE_FUNC": "openedxscorm.storage.s3",
+        "SCORM_S3_BUCKET_NAME": "<your-bucket-name>",
+        "SCORM_S3_QUERY_AUTH": False
```
We don't need to prefix these settings with "SCORM_" as they are already part of the "ScormXBlock" settings. So we should rename them to "S3_BUCKET_NAME"  and "S3_QUERY_AUTH".

regisb commented on README.rst:
```diff
@@ -82,12 +82,28 @@ By default, static assets are stored in the default Django storage backend. To o
     }
 
+This configuration is specifically for when using an S3 bucket to store SCORM assets.
+
```
minor formatting comment: 

```
* ``STORAGE_FUNC`` should be set to "openedxscorm.storage.s3".
* ``SCORM_S3_BUCKET_NAME`` should be replaced...
* ``SCORM_S3_QUERY_AUTH`` is a boolean flag...
...
```

regisb commented on openedxscorm/scormxblock.py:
```diff
@@ -202,6 +205,33 @@ def student_view(self, context=None):
+        -------
+        Response object containing the content of the requested file with the appropriate content type.
+        """
+        path = urlparse(_suffix).path
```
Since `_suffix` is used we should remove the "_" prefix from the variable name.

regisb commented on openedxscorm/scormxblock.py:
```diff
@@ -202,6 +205,33 @@ def student_view(self, context=None):
         return frag
 
+    @XBlock.handler
+    def assets_proxy(self, request, _suffix):
```
If I understand your implementation correctly, there is no need to go through the `assets_proxy` when authentication is disabled, right? Can you please add a corresponding comment in this docstring?

regisb commented on openedxscorm/storage.py:
```diff
@@ -0,0 +1,81 @@
+        Override url method of S3Boto3Storage
+        """
+        if not self.querystring_auth:
+            return self.generate_url(name, parameters, expire)
```
Please add the following comment:

```
# No need to use assets proxy when authentication is disabled
```

regisb commented on openedxscorm/storage.py:
```diff
@@ -0,0 +1,81 @@
+                handler_url += '/'
+
+            # construct the URL for proxy function
+            return f'{handler_url}{self.xblock.index_page_path}'
```
This piece of code can be shortened to:

```
handler_url = self.xblock.runtime.handler_url(self.xblock, 'assets_proxy').rstrip("?/")
return f"{handler_url}/{self.xblock.index_page_path}"
```

Also, shouldn't `self.xblock.index_path_path` be replaced by `name`?

regisb replied:
@naumanshafi's answer:

> **Question**: Also, shouldn't self.xblock.index_path_path be replaced by name?
> 
> - In the provided code, when the call comes from the index_page_url method and the bucket is private, the intention is to serve the first file (index page) using the assets_proxy function.
> 
> - In this case, instead of using `name`, it should use `self.xblock.index_page_path` to construct the URL for the proxy function.
> 
> - By using `self.xblock.index_page_path`, we ensure that the URL is constructed correctly based on the specific file requested, allowing the assets_proxy function to handle the serving of the file using the appropriate URL.

regisb replied:
Right, I understand. Could you please add a comment that to explain that the call comes from the `index_page_url` property?

regisb commented on openedxscorm/storage.py:
```diff
@@ -0,0 +1,81 @@
+            # construct the URL for proxy function
+            return f'{handler_url}{self.xblock.index_page_path}'
+
+        return self.generate_url(os.path.join(self.xblock.extract_folder_path, name), parameters, expire)
```
I'm not sure I understand in which cases we land here. Is it only when pointing to the legacy storage files in the `extract_folder_base_path`? If yes, we can ignore this edge case.

regisb replied:
@naumanshafi's answer:

> **Question**:I'm not sure I understand in which cases we land here. Is it only when pointing to the legacy storage files in the extract_folder_base_path? If yes, we can ignore this edge case.
> 
> - The line return `self.generate_url(os.path.join(self.xblock.extract_folder_path, name), parameters, expire)` is used when the call originates from the assets_proxy function.
> 
> - In this specific case when the bucket is Private, the assets_proxy function is responsible for serving assets by generating a pre-signed URL for accessing the content in the AWS S3 bucket. The URL is then used to retrieve the file content and return it as a response.
>

regisb commented on openedxscorm/scormxblock.py:
```diff
@@ -202,6 +205,33 @@ def student_view(self, context=None):
+        file_name = os.path.basename(path)
+        signed_url = self.storage.url(path)
+        file_type, _ = mimetypes.guess_type(file_name)
+        file_content = requests.get(signed_url).content
```
I'd rather avoid `requests` here and use the standard `urllib.urlopen` instead.

regisb commented on openedxscorm/storage.py:
```diff
@@ -0,0 +1,81 @@
+
+        return self.generate_url(os.path.join(self.xblock.extract_folder_path, name), parameters, expire)
+
+    def generate_url(self, name, parameters, expire):
```
What can't we just return `super().url(name, parameters, expire)` here?

regisb replied:
@naumanshafi's answer:

> **Question**:What can't we just return super().url(name, parameters, expire) here?
> 
> - Returning super().url(name, parameters, expire) would not be suitable in this scenario where the S3 bucket is private and a custom domain is set.
> 
> - When S3_CUSTOM_DOMAIN is configured with a custom domain for serving the files, the files in the bucket will not be directly accessible through the custom domain.
> 
> - In such cases, it is crucial to generate a pre-signed URL to provide temporary access credentials and permissions for accessing the private files. By generating a pre-signed URL, you ensure that authorized users can securely access the files in the private bucket, even when a custom domain is set.

naumanshafi commented:
**Question**:I'm not sure I understand in which cases we land here. Is it only when pointing to the legacy storage files in the extract_folder_base_path? If yes, we can ignore this edge case.

- The line return `self.generate_url(os.path.join(self.xblock.extract_folder_path, name), parameters, expire)` is used when the call originates from the assets_proxy function.

- In this specific case when the bucket is Private, the assets_proxy function is responsible for serving assets by generating a pre-signed URL for accessing the content in the AWS S3 bucket. The URL is then used to retrieve the file content and return it as a response.

naumanshafi commented:
**Question**: Also, shouldn't self.xblock.index_path_path be replaced by name?

- In the provided code, when the call comes from the index_page_url method and the bucket is private, the intention is to serve the first file (index page) using the assets_proxy function.

- In this case, instead of using `name`, it should use `self.xblock.index_page_path` to construct the URL for the proxy function.

- By using `self.xblock.index_page_path`, we ensure that the URL is constructed correctly based on the specific file requested, allowing the assets_proxy function to handle the serving of the file using the appropriate URL.

regisb commented:
In the future @naumanshafi please reply to each comment in its dedicated thread. If you put all your answers in the main conversation it makes it very difficult to track what was the initial question.

regisb reviewed (approved):
Thanks for your patience!
